README
NAME
    `Net::Async::Matrix' - use Matrix with IO::Async

SYNOPSIS
     use Net::Async::Matrix;
     use IO::Async::Loop;

     my $loop = IO::Async::Loop->new;

     my $matrix = Net::Async::Matrix->new(
        server => "my.home.server",
     );

     $loop->add( $matrix );

     $matrix->login(
        user     => '@my-user:home.server',
        password => 'SeKr1t',
     )->get;

DESCRIPTION
    Matrix is an new open standard for interoperable Instant Messaging and
    VoIP, providing pragmatic HTTP APIs and open source reference
    implementations for creating and running your own real-time
    communication infrastructure.

    This module allows an program to interact with a Matrix homeserver as a
    connected user client.

    http://matrix.org/

EVENTS
    The following events are invoked, either using subclass methods or
    `CODE' references in parameters:

  on_log $message
    A request to write a debugging log message. This is provided temporarily
    for development and debugging purposes, but will at some point be
    removed when the code has reached a certain level of stability.

  on_presence $user, %changes
    Invoked on receipt of a user presence change event from the homeserver.
    `%changes' will map user state field names to 2-element ARRAY
    references, each containing the old and new values of that field.

  on_room_new $room
    Invoked when a new room first becomes known about.

    Passed an instance of Net::Async::Matrix::Room.

  on_room_del $room
    Invoked when the user has now left a room.

  on_invite $event
    Invoked on receipt of a room invite. The `$event' will contain the plain
    Matrix event as received; with at least the keys `inviter' and
    `room_id'.

  on_unknown_event $event
    Invoked on receipt of any sort of event from the event stream, that is
    not recognised by any of the other code. This can be used to handle new
    kinds of incoming events.

PARAMETERS
    The following named parameters may be passed to `new' or `configure'. In
    addition, `CODE' references for event handlers using the event names
    listed above can also be given.

  server => STRING
    Hostname and port number to contact the homeserver at. Given in the form

     $hostname:$port

    This string will be interpolated directly into HTTP request URLs.

  SSL => BOOL
    Whether to use SSL/TLS to communicate with the homeserver. Defaults
    false.

  SSL_* => ...
    Any other parameters whose names begin `SSL_' will be stored for passing
    to the HTTP user agent. See IO::Socket::SSL for more detail.

  path_prefix => STRING
    Optional. Gives the path prefix to find the Matrix client API at.
    Normally this should not need modification.

  on_room_member, on_room_message => CODE
    Optional. Sets default event handlers on new room objects.

METHODS
    The following methods documented with a trailing call to `->get' return
    Future instances.

  $matrix->login( %params )->get
    Performs the necessary steps required to authenticate with the
    configured Home Server, actually obtain an access token and starting the
    event stream. The returned `Future' will eventually yield the `$matrix'
    object itself, so it can be easily chained.

    There are various methods of logging in supported by Matrix; the
    following sets of arguments determine which is used:

    user_id, password
        Log in via the `m.login.password' method.

    user_id, access_token
        Directly sets the `user_id' and `access_token' fields, bypassing the
        usual login semantics. This presumes you already have an existing
        access token to re-use, obtained by some other mechanism. This
        exists largely for testing purposes.

  $matrix->register( %params )->get
    Performs the necessary steps required to create a new account on the
    configured Home Server.

  $f = $matrix->start
    Performs the initial IMSync on the server, and starts the event stream
    to begin receiving events.

    While this method does return a `Future' it is not required that the
    caller keep track of this; the object itself will store it. It will
    complete when the initial IMSync has fininshed, and the event stream has
    started.

    If the initial sync has already been requested, this method simply
    returns the future it returned the last time, ensuring that you can
    await the client starting up simply by calling it; it will not start a
    second time.

  $matrix->stop
    Stops the event stream. After calling this you will need to use `start'
    again to continue receiving events.

  $user = $matrix->myself
    Returns the user object representing the connected user.

  $user = $matrix->user( $user_id )
    Returns the user object representing a user of the given ID, if defined,
    or `undef'.

  $name = $matrix->get_displayname->get
  $matrix->set_displayname( $name )->get
    Accessor and mutator for the user account's "display name" profile
    field.

  ( $presence, $msg ) = $matrix->get_presence->get
  $matrix->set_presence( $presence, $msg )->get
    Accessor and mutator for the user's current presence state and optional
    status message string.

  ( $room, $room_alias ) = $matrix->create_room( $alias_localpart )->get
    Requests the creation of a new room and associates a new alias with the
    given localpart on the server. The returned `Future' will return an
    instance of Net::Async::Matrix::Room and a string containing the full
    alias that was created.

  $room = $matrix->join_room( $room_alias_or_id )->get
    Requests to join an existing room with the given alias name or plain
    room ID. If this room is already known by the `$matrix' object, this
    method simply returns it.

  $matrix->add_alias( $alias, $room_id )->get
  $matrix->delete_alias( $alias )->get
    Performs a directory server request to create the given room alias name,
    to point at the room ID, or to remove it again.

    Note that this is likely only to be supported for alias names scoped
    within the homeserver the client is connected to, and that additionally
    some form of permissions system may be in effect on the server to limit
    access to the directory server.

USER STRUCTURES
    Parameters documented as `$user' receive a user struct, which supports
    the following methods:

  $user_id = $user->user_id
    User ID of the user.

  $displayname = $user->displayname
    Profile displayname of the user.

  $presence = $user->presence
    Presence state. One of `offline', `unavailable' or `online'.

  $last_active = $user->last_active
    Epoch time that the user was last active.

SUBCLASSING METHODS
    The following methods are not normally required by users of this class,
    but are provided for the convenience of subclasses to override.

  $room = $matrix->make_room( %params )
    Returns a new instance of Net::Async::Matrix::Room.

SEE ALSO
    *   http://matrix.org/ - matrix.org home page

    *   https://github.com/matrix-org - matrix.org on github

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Net-Async-Matrix

0.13    2014/12/10 17:08:42
        [CHANGES]
         * Initial start at some unit tests

        [BUGFIXES]
         * 'use URI' since it's needed
         * Clear the (global and room) initialSync futures if the operation
           fails so that the next call will re-attempt it

0.12    2014/12/02 15:50:18
        [CHANGES]
         * Cope with new m.room.power_levels event format
         * Allow applications to handle new/unrecognised events by adding
           'on_unknown_events' event
        
        [BUGFIXES]
         * Don't emit rich formatted messages if there are no actual
           formatting tags set on the String::Tagged instance

0.11    2014/11/14 18:25:06
        [CHANGES]
         * Support formatted messages in the "org.matrix.custom.html" format
           by conversion to/from String::Tagged::Formatting instances
         * Display formatted messages and image URLs in the client

0.10    2014/11/04 16:15:31
        [CHANGES]
         * Extracted 'make_room' as a subclassable method
         * Use new Net::Async::HTTP 0.36 ability to store SSL parameters
         * Pass $event as another param in on{_back,}_message so that new
           timestamp API is visible

        [BUGFIXES]
         * Declare correct version of IaNotifier in Build.PL
         * Not all '/register' alternatives will have 'stages'
         * Fix to ->join_room ensuring that $room itself is always returned

0.09    2014/10/07 18:05:58
        [CHANGES]
         * Many more events are handled and reported via on_state:
            + m.room.aliases
            + m.room.join_rules
            + m.room.*_level
         * Added ability to send invites and an event for incoming invites
         * Room membership change events now also include the member making
           the change as well as the member the change is about
         * Allow setting room names and topics, adding and removing aliases,
           and changing member and room power levels

        [BUGFIXES]
         * Fetch full room initial state on join or create before invoking
           on_state_synced event

0.08    2014/09/24 17:09:50
        [CHANGES]
         * Added $room->joined_members, to allow filtering of other states
           (invitees, kicks/bans)
         * Fix ->register for new Registration API, including multi-stage
           flows
         * Added $room->leave
         * Client improvements
            + per-room commands: /me and /leave
            + display a user total count on the user list for a room

0.07    2014/09/22 17:05:01
        [CHANGES]
         * Support /register Matrix API with user_id + password
         * Fix on_membership event argument order to match on_state_change

0.06    2014/09/19 17:35:40
        [CHANGES]
         * New room event model to handle back-pagination as well as forward
           events
         * Support room names and topics
         * Much improvement to 'client.pl' - nicer formatting, room topics,
           put member joins/leaves/renames in the event log
         * Support /login Matrix API with user_id + password

0.05    2014/09/10 11:25:22
        [CHANGES]
         * Split room 'on_member' events into separate events for membership
           vs. presence
         * Various neatenings and improvements in included 'client.pl' script

        [BUGFIXES]
         * Apply a timeout for longpoll so we don't end up waiting forever
         * Invoke ->log method of containing $matrix object, not toplevel
           main::log()

0.04    2014/09/04 18:10:00
        [CHANGES]
         * Many and various improvements to the display of the sample client
         * Added 'on_synced_*' events to Room objects, removed 'on_room_synced'
           from Matrix object

0.03    2014/09/03 16:30:54
        [CHANGES]
         * Many more updates for current Matrix API
         * Split 'on_room_add' into on_room_new + on_room_synced events
         * Support SSL

0.02    2014/08/27 19:30:34
        [CHANGES]
         * Updates for latest Matrix REST API
         * ->create_room to return a new Room object
         * Document the user and membership structs
         * Initial sync of rooms when joining
         * Don't store room alias names
         * Added $room->members
         * Allow setting default on_room_* event handlers

        [BUGFIXES]
         * Remember to fill in request content body of event stream requests,
           so wrappers around NaHTTP still see it

0.01    2014/08/23 10:29:11
        First version, released on an unsuspecting world.



lib/Net/Async/Matrix/Utils.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Net::Async::Matrix::Utils;

use strict;
use warnings;

our $VERSION = '0.13';
$VERSION = eval $VERSION;

use Exporter 'import';
our @EXPORT_OK = qw(
   parse_formatted_message
   build_formatted_message
);

use String::Tagged 0.11;

# Optionally parse HTML rich-formatted body; but don't get too upset if we
# don't have these installed
use constant CAN_PARSE_HTML => eval {
   require HTML::TreeBuilder;
   require Convert::Color::HTML;
};

# Optionally build HTML rich-formatted body; but don't get too upset if we
# don't have this installed
use constant CAN_BUILD_HTML => eval {
   require String::Tagged::HTML;
   require Convert::Color::HTML;
};

=head1 NAME

C<Net::Async::Matrix::Utils> - support utilities for L<Net::Async::Matrix>

=head1 DESCRIPTION

=cut

=head1 FUNCTIONS

=cut

=head2 $st = parse_formatted_message( $content )

Given the content of a C<m.room.message> event of C<m.text> or C<m.emote>
type, returns a L<String::Tagged> instance containing the text of the message
with formatting in L<String::Tagged::Formatting> style. If the message is not
formatted, or the formatting is of a kind not recognised, the plain-text body
is returned in an instance with no tags.

The following formats are recognised:

=over 4

=item org.matrix.custom.html

This format requires the presence of L<HTML::TreeBuilder> to parse; it will be
ignored if this module is not available.

=back

=cut

sub parse_formatted_message
{
   my ( $content ) = @_;

   for my $format ( $content->{format} ) {
      last if !$format;

      return _parse_html_body( $content->{formatted_body} ) if
         CAN_PARSE_HTML and $format eq "org.matrix.custom.html";
   }

   return String::Tagged->new( $content->{body} );
}

sub _parse_html_body
{
   my ( $formatted ) = @_;

   return _traverse_html( HTML::TreeBuilder->new_from_content( $formatted )
      ->find_by_tag_name( 'body' )
   );
}

sub _traverse_html
{
   my ( $node ) = @_;

   # Plain text
   return String::Tagged->new( $node ) if !ref $node;

   my %tags;
   for ( $node->tag ) {
      $_ eq "b" and $tags{bold}++,      last;
      $_ eq "i" and $tags{italic}++,    last;
      $_ eq "u" and $tags{underline}++, last;

      if( $_ eq "font" ) {
         my %attrs = $node->all_attr;

         my $fg = defined $attrs{color} ?
            eval { Convert::Color::HTML->new( $attrs{color} ) } : 
            undef;

         $tags{fg} = $fg if defined $fg;

         last;
      }
   }

   my $ret = String::Tagged->new;

   $ret .= _traverse_html( $_ ) for $node->content_list;

   $ret->apply_tag( 0, length $ret, $_, $tags{$_} ) for keys %tags;

   return $ret;
}

=head2 $content = build_formatted_message( $str )

Given a L<String::Tagged::Formatting> instance or plain string, returns a
content HASH reference encoding the formatting the message. Plain strings are
returned simply as a plain-text body; formatted instances will be output as
formatted content if possible:

=over 4

=item org.matrix.custom.html

This format is output for formatted messages if L<String::Tagged::HTML> is
available.

=back

=cut

sub build_formatted_message
{
   my ( $str ) = @_;

   return { body => $str } if !ref $str;
   return { body => $str->str } if !$str->tagnames;

   if( CAN_BUILD_HTML ) {
      my $html = String::Tagged::HTML->clone( $str,
         only_tags => [qw( bold under italic strike fg )],
         convert_tags => {
            bold   => "b",
            under  => "u",
            italic => "i",
            strike => "strike",
            fg     => sub { font => { color => $_[1]->as_html->name } },
         },
      );

      return {
         body => $str->str,
         format => "org.matrix.custom.html",
         formatted_body => $html->as_html,
      };
   }
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


Build.PL
use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
   module_name => 'Net::Async::Matrix',
   requires => {
      'Data::Dump' => 0,
      'Future::Utils' => 0,
      'IO::Async::Notifier' => '0.63',
      'JSON' => 0,
      'List::Util' => '1.29',
      'Net::Async::HTTP' => '0.36',
      'POSIX' => 0,
      'String::Tagged' => '0.11',
      'Struct::Dumb' => 0,
      'URI' => 0,
   },
   recommends => {
      'Digest::HMAC_MD5' => 0,

      # For formatted message support
      'Convert::Color::HTML' => 0,
      'HTML::TreeBuilder' => '5',
      'String::Tagged::HTML' => 0,
   },
   test_requires => {
      'Test::More' => '0.88', # done_testing
      'Test::Async::HTTP' => '0.02',
   },
   auto_configure_requires => 0, # Don't add M::B to configure_requires
   license => 'perl',
   create_makefile_pl => 'small',
   create_license => 1,
   create_readme  => 1,
   meta_merge => {
      resources => {
         x_IRC => "irc://irc.perl.org/#io-async",
      },
   },
);

$build->create_build_script;
